bot/exts/fun/typst.py

```diff
@@ -0,0 +1,309 @@
+    """Represents an error caused by invalid typst source code."""
+
+    def __init__(self, logs: str | None):
+        super().__init__(logs)
+        self.logs = logs
+
+
+class TypstTimeoutError(Exception):
+    """Represents an error caused by the Typst rendering taking too long."""
+
+
+class OutputTooBigError(Exception):
+    """Represents an error caused by the Typst output image being too big."""
+
+
+class EmptyImageError(Exception):
+    """Represents an error caused by the output image being empty."""
+
+
+class TypstWorkerCrashedError(Exception):
+    """Represents an error caused by Typst rendering process crashing. This can mean the memory limit was exceeded."""
+
+
+class Typst(commands.Cog):
+    """Renders typst."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    async def _setup_typst(self) -> None:
+        await self._ensure_typst_executable()
+        await self._setup_packages()
+
+    async def _ensure_typst_executable(self) -> None:
+        path = Path(Config.typst_path).resolve()
+        if path.exists():
+            if not path.is_file():
+                raise ValueError("Typst path exists but doesn't point to a file:", path)
+        else:
+            log.info("Typst executable not found at '%s', downloading", path)
+            await self._download_typst_executable()
+        proc = await asyncio.subprocess.create_subprocess_exec(
+            path,
+            "--version",
+            stdin=asyncio.subprocess.DEVNULL,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+        )
+        stdout, _ = await proc.communicate()
+        log.info(
+            f"Typst executable reports itself as {stdout.decode('utf-8').strip()!r}."
+        )
+
+    async def _download_typst_executable(self) -> None:
+        if not Config.typst_archive_url:
+            raise ValueError("Trying to download Typst but the archive URL isn't set")
+        if not Config.typst_archive_sha256:
+            raise ValueError("Trying to download Typst but the archive hash isn't set")
+        async with self.bot.http_session.get(
+            Config.typst_archive_url, raise_for_status=True
+        ) as response:
+            arc_data = await response.read()
+        digest = hashlib.sha256(arc_data).hexdigest()
+        if digest != Config.typst_archive_sha256:
+            raise ValueError(
+                f"Retrieved archive doesn't match hash {Config.typst_archive_sha256}; "
+                f"instead got file with size {len(arc_data)} and hash {digest}"
+            )
+        log.info("Retrieved Typst archive, unpacking")
+        typst_executable = archive_retrieve_file(
+            arc_data,
+            filename="typst.exe" if sys.platform == "win32" else "typst",
+        )
+        log.info("Storing the typst executable on disk")
+        exe_path = Path(Config.typst_path)
+        exe_path.parent.mkdir(exist_ok=True)
+        exe_path.write_bytes(typst_executable)
```

Comment:
This isn't going to work in our infrastructure. We use read-only file systems.

I am also very uneasy with the fact we're downloading and executing a binary from the internet like this into the bot's pod.

Is there a typst API similar to what we use for latex? Either that, or we can look into having a typst deployment in our infra, where it won't have access to the bot's secrets.